Adding function to convert Module to channels last

VitalyFedyunin opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #31139 [WIP] Fix cudnn channels_last descriptio problem
* #31131 Add memory_format support to `zeros`, `ones`, `full` (still need tests)
* #30089 Switch default memory format of clone operator to Preserve
* #30088 Switch default memory format of to (and similar) operators to Preserve
* #30087 Switch default memory format of _like operators to Preserve
* **#28991 Adding function to convert Module to channels last**

Differential Revision: [D18430810](https://our.internmc.facebook.com/intern/diff/D18430810)

VitalyFedyunin committed:
Adding function to convert Module to channels last

[ghstack-poisoned]

soumith commented on torch/nn/modules/module.py:
```diff
@@ -288,6 +288,13 @@ def apply(self, fn):
         fn(self)
         return self
 
+    def channels_last(self):
```
this seems like an overkill. can you please just add in `.to`.

`model.to(memory_format=torch.channels_last)`

VitalyFedyunin replied:
Absolutely

VitalyFedyunin replied:
Done

VitalyFedyunin committed:
add TensroIterator tests on "Adding function to convert Module to channels last"




[ghstack-poisoned]

VitalyFedyunin committed:
Add some tests on "Adding function to convert Module to channels last"




[ghstack-poisoned]

VitalyFedyunin committed:
Addressed PR feedback on "Adding function to convert Module to channels last"




[ghstack-poisoned]

VitalyFedyunin committed:
Changed model conversion code on "Adding function to convert Module to channels last"




[ghstack-poisoned]

VitalyFedyunin committed:
Linter fix on "Adding function to convert Module to channels last"




[ghstack-poisoned]

VitalyFedyunin committed:
Fix resize_ on "Adding function to convert Module to channels last"




[ghstack-poisoned]

VitalyFedyunin committed:
Fix TensorIterator on "Adding function to convert Module to channels last"




[ghstack-poisoned]

VitalyFedyunin committed:
Fix Autograd on "Adding function to convert Module to channels last"


Differential Revision: [D18430810](https://our.internmc.facebook.com/intern/diff/D18430810)

[ghstack-poisoned]

gchanan commented on torch/nn/modules/module.py:
```diff
@@ -419,8 +424,13 @@ def to(self, *args, **kwargs):
                 raise TypeError('nn.Module.to only accepts floating point '
                                 'dtypes, but got desired dtype={}'.format(dtype))
 
+        convert_to_format = kwargs.get('memory_format', None)
```
is there a reason you didn't modify _parse_to?  It seems nice to keep the parsing logic in one place.

gchanan commented on torch/nn/modules/module.py:
```diff
@@ -419,8 +424,13 @@ def to(self, *args, **kwargs):
-            return t.to(device, dtype if t.is_floating_point() else None, non_blocking)
+            result = t
+            if convert_to_format is not None and t.dim() == 4:
+                result = t.contiguous(memory_format=convert_to_format)
```
if you can't call `to(memory_format=...)` for now that seems fine, just comment it.

VitalyFedyunin replied:
done

VitalyFedyunin committed:
Fix resize_as_ with new test on "Adding function to convert Module to channels last"


Differential Revision: [D18430810](https://our.internmc.facebook.com/intern/diff/D18430810)

[ghstack-poisoned]

VitalyFedyunin committed:
Fix resize_as_ on "Adding function to convert Module to channels last"


Differential Revision: [D18430810](https://our.internmc.facebook.com/intern/diff/D18430810)

[ghstack-poisoned]

VitalyFedyunin committed:
Fix resize_ on "Adding function to convert Module to channels last"


Differential Revision: [D18430810](https://our.internmc.facebook.com/intern/diff/D18430810)

[ghstack-poisoned]

VitalyFedyunin committed:
Fix model.to on "Adding function to convert Module to channels last"


Differential Revision: [D18430810](https://our.internmc.facebook.com/intern/diff/D18430810)

[ghstack-poisoned]

gchanan commented on torch/nn/modules/module.py:
```diff
@@ -378,6 +380,9 @@ def to(self, *args, **kwargs):
             tensor (torch.Tensor): Tensor whose dtype and device are the desired
                 dtype and device for all parameters and buffers in this module
+            memory_format (:class:`torch.memory_format`): the desired memory
+                format  for all parametes and buffers in this module (keyword
```
spelling: parameters

gchanan replied:
this doesn't seem true (I guess it depends on your definition of "desired").  But this only converts the 4-dimensional parameters, right?

gchanan commented on torch/nn/modules/module.py:
```diff
@@ -419,8 +424,16 @@ def to(self, *args, **kwargs):
                 raise TypeError('nn.Module.to only accepts floating point '
                                 'dtypes, but got desired dtype={}'.format(dtype))
 
+        # memory_format handled separately as we can't call `to` operator with
```
I don't really understand the comment.  Why can't you call to with only memory format?  Is it because it's not bindable yet? (list an issue here).  Or it's not part of the API (describe why?)

VitalyFedyunin replied:
Fixed.

gchanan commented on torch/nn/modules/module.py:
```diff
@@ -419,8 +424,16 @@ def to(self, *args, **kwargs):
                                 'dtypes, but got desired dtype={}'.format(dtype))
 
+        # memory_format handled separately as we can't call `to` operator with
+        # only memory_format, and we want to keep ability to convert layouts,
```
I don't really understand this clause, don't we already have the ability to convert layouts, etc. without converting dtype or device?

Also, what happens if you call `to(tensor)`?  Does it convert the memory format?  My overall point is I still don't see why we don't just do the parsing correctly in _parse_to -- as this is written, I can do something like `to(tensor, memory_format=...)` but that's not the intention of the API.  But that's perfectly expressible with the python_arg_parser.

VitalyFedyunin replied:
Fixed.

VitalyFedyunin committed:
Changes in module.to on "Adding function to convert Module to channels last"


Differential Revision: [D18430810](https://our.internmc.facebook.com/intern/diff/D18430810)

[ghstack-poisoned]

VitalyFedyunin committed:
resize_ PR fixes on "Adding function to convert Module to channels last"


Differential Revision: [D18430810](https://our.internmc.facebook.com/intern/diff/D18430810)

[ghstack-poisoned]

VitalyFedyunin committed:
Add resize_as_ docs on "Adding function to convert Module to channels last"


Differential Revision: [D18430810](https://our.internmc.facebook.com/intern/diff/D18430810)

[ghstack-poisoned]

VitalyFedyunin commented:
All comments addressed

VitalyFedyunin committed:
fix resize_as_ docs on "Adding function to convert Module to channels last"


Differential Revision: [D18430810](https://our.internmc.facebook.com/intern/diff/D18430810)

[ghstack-poisoned]

VitalyFedyunin committed:
fix resize_as_ docs on "Adding function to convert Module to channels last"


Differential Revision: [D18430810](https://our.internmc.facebook.com/intern/diff/D18430810)

[ghstack-poisoned]

VitalyFedyunin committed:
rebase to master on "Adding function to convert Module to channels last"


Differential Revision: [D18430810](https://our.internmc.facebook.com/intern/diff/D18430810)

[ghstack-poisoned]

VitalyFedyunin commented:
CC @soumith

VitalyFedyunin committed:
rebase to master + fix tests on "Adding function to convert Module to channels last"


Differential Revision: [D18430810](https://our.internmc.facebook.com/intern/diff/D18430810)

[ghstack-poisoned]

VitalyFedyunin committed:
fix tests code on "Adding function to convert Module to channels last"


Differential Revision: [D18430810](https://our.internmc.facebook.com/intern/diff/D18430810)

[ghstack-poisoned]

soumith reviewed (approved):
lgtm from my request. cc: @gchanan as the other reviewer who requested changes,

VitalyFedyunin committed:
add experimental code to check model on "Adding function to convert Module to channels last"


Differential Revision: [D18430810](https://our.internmc.facebook.com/intern/diff/D18430810)

[ghstack-poisoned]

VitalyFedyunin committed:
experiments with tag on "Adding function to convert Module to channels last"


Differential Revision: [D18430810](https://our.internmc.facebook.com/intern/diff/D18430810)

[ghstack-poisoned]

VitalyFedyunin committed:
rebase to master on "Adding function to convert Module to channels last"


Differential Revision: [D18430810](https://our.internmc.facebook.com/intern/diff/D18430810)

[ghstack-poisoned]

kostmo commented:
## CircleCI build failures summary
As of commit ea3ed3d:
* **1/1** broken upstream at merge base 56de885 (see [grid view](https://dr.pytorch.org/master-timeline.html?count=1&sha1=56de8853dad19a050bc9ce5f89c3f7572ca8a15c&should_suppress_scheduled_builds=true&should_suppress_fully_successful_columns=true&max_columns_suppress_successful=35))
    * You may want to rebase on the `viable/strict` branch (see its [recency history](https://dr.pytorch.org/master-viable-commits.html)):
        * If your commit is newer than `viable/strict`, you can try basing on an older, stable commit:
          ```
          git fetch viable/strict
          git rebase --onto viable/strict $(git merge-base origin/master HEAD)
          ```
        * If your commit is older than `viable/strict`:
          ```
          git fetch viable/strict
          git rebase viable/strict
          ```
* **0/1** failures introduced in this PR



## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=ea3ed3d3207a79a1eae377fb310205efb890b22b).

### 1 failure *not* recognized by patterns:
| Job | Step |
| --- | --- |
| ![CircleCI](https://avatars0.githubusercontent.com/ml/7?s=12) <sup>pytorch_windows_build</sup> | <sup>Build</sup> |


---
This comment was automatically generated by [Dr. CI](https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment).
Follow [this link to opt-out](https://dr.pytorch.org/admin/comments-opt-out.html) of these comments for your Pull Requests.


Please report bugs/suggestions on the [GitHub issue tracker](https://github.com/kostmo/circleci-failure-tracker/issues).

gchanan commented on torch/nn/modules/module.py:
```diff
@@ -412,14 +417,16 @@ def to(self, *args, **kwargs):
 
         def convert(t):
+            if convert_to_format is not None and t.dim() == 4:
+                return t.to(device, dtype if t.is_floating_point() else None, non_blocking, memory_format=convert_to_format)
```
this doesn't match the documentation (which says the only case for `to` with memory-format is the 1-arg case).

VitalyFedyunin replied:
Pardon, but I read 'This can be called as' as: here examples of calls, but they are not limited to this options.

gchanan replied:
that's not my reading of it, although I can see why you read it that way.

In particular, before memory_format was introduced, it corresponded exactly to the function signatures:
https://github.com/pytorch/pytorch/blob/66f2bba8527d941c7d73d3cae9e9576c601587a6/torch/csrc/autograd/utils/python_arg_parsing.h#L15-L17

(remove copy because it's not supported and remove memory_format because we are considering the case before memory_format was introduced).

So, the introduction of memory_format changed this from "these are the supported signatures" to "these are some examples of supported signatures".  I think the former is more useful and we should change it back.

gchanan replied:
And actually, the example you added (`.. function:: to(memory_format=torch.channels_last)`) doesn't work because the parsing code hasn't been updated, right?

gchanan replied:
IMO we should do the following:
1) Add a memory_format only-overload to the python parsing.
2) List the valid calls, i.e.:
```
        .. function:: to(device=None, dtype=None, non_blocking=False, memory_format=None)
        .. function:: to(dtype, non_blocking=False, memory_format=None)
        .. function:: to(tensor, non_blocking=False, memory_format=None)
        .. function:: to(memory_format)
```
(or similar).

VitalyFedyunin replied:
I had parsing updated https://github.com/pytorch/pytorch/pull/28991/files#diff-
7a05cfd8eb442889dffd6c3d2e4d0ddcR24 

Will update inline and html docs as follow-up PR

gchanan commented on torch/nn/modules/module.py:
```diff
@@ -378,6 +380,9 @@ def to(self, *args, **kwargs):
                 the floating point parameters and buffers in this module
             tensor (torch.Tensor): Tensor whose dtype and device are the desired
                 dtype and device for all parameters and buffers in this module
+            memory_format (:class:`torch.memory_format`): the desired memory
```
are you planning to have a reference section for memory_format that you can point to?  This description isn't full enough for the long term.  (e.g. why only 4D parameters/buffers? -- it's not clear)

VitalyFedyunin replied:
Yes as soon as we land new defaults for .clone .to *_like ops I will work on updating docs.

ppwwyyxx commented:
This PR adds an API `model.to(memory_layout=torch.channels_last)`, which recursively applies NHWC layout to all 4D parameters in the model.

I think this is making an assumption that "any 4D parameters in the module needs a conversion to NHWC layout, if user wants to use nvidia's NHWC kernels". From what I can see this assumption is limiting in many ways:
* Not all 4D parameters in a user's custom model means "NCHW". The dimensions could have arbitrary meanings, and changing their memory layout may hurt performance.
* Not all parameters that are in "NCHW" format needs to be changed to a different format. They need to be changed only in certain layers where Nvidia provides an implementation, e.g. convolution. Imagine a user's custom variant of convolution kernel that may or may not benefit from this layout change.
* Converting the layout of filters may not be the best strategy. Converting filters from NCHW to NHWC may happen to be what Nvidia now asks for when using NHWC inputs. However it seems possible that the next version of GPUs, or some other runtime library have different requirements. They may for example decide that NHWC inputs works best together with NCHW filters. Aligning the layout of filters together with the layout of inputs, under the same API, doesn't seem like a good idea.

It may be better if such an API:
* only change layout of weights from certain fixed set of layers that can benefit from this change, e.g. convolution
* has a more backend-specific name, such as `model = optimize_for_cudnn_channels_last(model)`:
  * It seems to me that having to use NHWC filter layout is merely an optimization specifically made for cudnn, instead of a generally applied strategy.
  * what filter layout is best for NHWC input layout is a detail of cudnn, and should not be exposed to users. The API `model.to(memory_layout=)` implies that it is changing the layout of weights to NHWC, which is a cudnn requirement that might change in future hardwares.
  * Calling this function should be optional (not sure if it is now). Using NCHW filter layout together with NHWC input should be allowed with a perf penalty on users. With `model.to(memory_layout=)` it gives an impression that it is required (just like `to(dtype)` or `to(device)` has to match)

dreiss commented:
@ppwwyyxx , I totally agree.
